### PR TITLE
Update PETSc to 3.19 series

### PR DESCRIPTION
**Pull request by cticenhour**
## Reason
MOOSE is currently using the PETSc 3.16 series, while PETSc itself is on the 3.19 series. The PETSc submodule in MOOSE should be updated to the new release. 

## Design
Update PETSc, and fixup any broken MOOSE tests or systems.

## Impact
Shouldn't change core prior functionality of MOOSE, but provides users with new features and bugfixes from the new release.

### CODE REVIEW

**Review by loganharbour — CHANGES_REQUESTED** [review]
Need to update https://github.com/idaholab/moose/blob/5f86f198a56ed6985d1ebc282b81da3f049af20b/docker_ci/Dockerfile#L102

File: modules/doc/content/newsletter/2023/2023_08.md [hunk 1]
```diff
@@ -11,4 +11,31 @@ for a complete description of all MOOSE changes.
...
+```
+
+The following package versions and builds outlined in the update prompt should be
+at least the following:
+
+```
+Package                  Version                     Build
+============================================================
+moose-dev                2023.06.23                  build_1
+moose-libmesh            2023.06.23                  build_1
```
Comment (cticenhour) [inline modules/doc/content/newsletter/2023/2023_08.md:35]:
```suggestion
moose-dev                2023.07.31                  build_0
moose-libmesh            2023.07.26                  build_1
```

Reminder to do this before merging.

**Comment by cticenhour** [comment]
@lindsayad Your commit in #25166 has been moved over as [b58624e](https://github.com/idaholab/moose/pull/23502/commits/b58624ebeb7a3083dd585488f3cebc3e55161ea4). 

Working through app patches currently. RELAP-7 is done pending this PR going in. Griffin, SAM, and Subchannel are on-going.

**Comment by cticenhour** [comment]
I invalidated just in case it was something errant in the environment, but the `zstd` library it noted in the failure is distinct from `zlib` I thought - both compress/decompress algorithms, but different libraries. The change you made to `configure_petsc.sh` is only used when HDF5 needs to be built, which it shouldn't be in the case....we get it from conda here.

**Comment by lindsayad** [comment]
So what I'm hearing is that I'm probably not the problem in this case...? 😸

**Comment by cticenhour** [comment]
Doesn't seem so :) the invalidate did the trick in this case!

**Comment by YaqiWang** [comment]
Any help for fixing Griffin? All failing tests seem close related.

**Comment by cticenhour** [comment]
@YaqiWang if you'd like to help out with Griffin I would welcome it! I haven't been able to get to it quite yet.

**Comment by cticenhour** [comment]
@grmnptr and @zachmprince -- Pinging you both for help if you have some time. I was working this weekend on finishing off the last failing MOOSE test in stochastic tools heavy, and discovered *what* has changed in the PETSc TaO interface to cause the test failure, but I'm at a loss as to what the proper fix is (mostly due to unfamiliarity with TaO and the ST module itself). To give some background, the failing test is `examples/surrogates/gaussian_process.4D_HC_avg/train`, which seems (initially) to train well, but fails during the test of the calculated length factors. 

I initially bisected this back to [this PETSc commit in 3.18.0](https://gitlab.com/petsc/petsc/-/commit/9203fd1f6b645bbbc6de9e81e30dd3078b115bfc), which ensures consistency in the TaO linesearch `stepmax` and `stepmin` for every call to `TaoLineSearchApply_MT` by backing them up prior to the calculation of a bounded linesearch step and resetting them before finishing the computations. This wasn't sufficient to explain the differing behavior in the failing test, so I bisected again to find [this commit in 3.19.0](https://gitlab.com/petsc/petsc/-/commit/def90ec86b14b00b9b40b25f336eecd3d40eae8c), adding a failure point in this same section of code, which was already failing to converge (I will attach some TaO output in a dropdown section of my post below). Reverting both of these changes does indeed allow the test pair to pass successfully. 

My attempts to fix this involves changes to `stepmax` and `stepmin` to attempt to keep them away from a zero step, as well as to keep them within a prior hard-coded maximum dictated by the original commit I highlighted above. Tolerance adjustments also did not seem to have a demonstrable impact. Attempting other combinations of linesearch objects and CG types was also unsuccessful. Also one can see in the test CSV output that the training simulation consistently hits the rails of the `tuning_min` and `tuning_max` parameters in at least some of the calculated length factors. 

Take a look at the dropdown below to see some relevant output from the training test that is not performing correctly. I would really appreciate any help that either of you can provide!

<details>
<summary>TaO output using `show_optimization_details = true` and `tao_monitor`:</summary>
<br>
Details GP_normal_mc.i output for PR PETSc submodule:

```
   0 TAO,  Function value: 709.469,  Residual: 0.253798
   0 TAO,  Function value: 709.469,  Residual: 0.139531
   1 TAO,  Function value: 634.068,  Residual: 234.16
 Tao Object: 1 MPI process
   type: bncg
     CG Type: pcgd
     Skipped Stepdirection Updates: 0
     Scaled gradient steps: 1
     Pure gradient steps: 2
     Not a descent direction: 0
     Line search fails: 3
   TaoLineSearch Object: 1 MPI process
     type: more-thuente
     maximum function evaluations=30
     tolerances: ftol=0.0001, rtol=1e-10, gtol=0.9
     total number of function evaluations=0
     total number of gradient evaluations=0
     total number of function/gradient evaluations=1
     using variable bounds
     Termination reason: 6
   Active Set subset type: subvec
   convergence tolerances: gatol=1e-08,   steptol=0.,   gttol=0.
   Residual in Function/Gradient:=234.16
   Objective value=634.068
   total number of iterations=2,                          (max: 2000)
   total number of function/gradient evaluations=12,      (max: 4000)
   Solver terminated: -6   Line Search Failure
 Size	global =  5		local =  5
 #	Value
 0	1.18696
 1	0.680053
 2	0.206921
 3	0.749758
 4	0.379848
 ```

Details GP_normal_mc.i output with reverted changes:

```
   0 TAO,  Function value: 709.469,  Residual: 0.253798
   0 TAO,  Function value: 709.469,  Residual: 0.139531
   1 TAO,  Function value: 634.068,  Residual: 234.16
   2 TAO,  Function value: 427.762,  Residual: 292.089
 Tao Object: 1 MPI process
   type: bncg
     CG Type: pcgd
     Skipped Stepdirection Updates: 0
     Scaled gradient steps: 1
     Pure gradient steps: 2
     Not a descent direction: 0
     Line search fails: 3
   TaoLineSearch Object: 1 MPI process
     type: more-thuente
     maximum function evaluations=30
     tolerances: ftol=0.0001, rtol=1e-10, gtol=0.9
     total number of function evaluations=0
     total number of gradient evaluations=0
     total number of function/gradient evaluations=2
     using variable bounds
     Termination reason: 3
   Active Set subset type: subvec
   convergence tolerances: gatol=1e-08,   steptol=0.,   gttol=0.
   Residual in Function/Gradient:=292.089
   Objective value=427.762
   total number of iterations=3,                          (max: 2000)
   total number of function/gradient evaluations=19,      (max: 4000)
   Solver terminated: -6   Line Search Failure
 Size	global =  5		local =  5
 #	Value
 0	1.18032
 1	0.232354
 2	0.654185
 3	0.897953
 4	0.696113
 ```
 
</details>

**Comment by grmnptr** [comment]
I can take a look at this tomorrow!

**Comment by grmnptr** [comment]
Looks like the iteration values are the same, the new version just terminates one iteration early. Hmmm, I wonder if this is tolerance related.

**Comment by grmnptr** [comment]
Yeah, convergence result is:
(3) for the old one which is hitting the `gatol`
(6) for the now one which is hotting the `steptol`
Not sure how to change these, but we will see tomorrow

**Comment by cticenhour** [comment]
@YaqiWang were you able to make progress with a Griffin patch? With the closing of #25360, we only have app patches to go here.

**Comment by cticenhour** [comment]
@lindsayad will need to rebase this tomorrow in order to get an up-to-date CI result.

**Comment by cticenhour** [comment]
@lindsayad Subchannel patch is [ncrc/subchannel#294](https://github.inl.gov/ncrc/subchannel/pull/294), and is ready for a submodule update when we are ready to merge this. Is there a SAM patch ready?

**Comment by lindsayad** [comment]
No I will have to look at SAM

File: scripts/configure_petsc.sh [hunk 2]
```diff
@@ -69,9 +69,11 @@ function configure_petsc()
 
   # If HDF5 is not found locally, download it via PETSc
   HDF5_FORTRAN_STR=''
+  HDF5_CONFIGURE_STR=''
   if [ -z "$HDF5_STR" ]; then
     HDF5_STR='--download-hdf5=1'
     HDF5_FORTRAN_STR='--download-hdf5-fortran-bindings=0'
+    HDF5_CONFIGURE_STR='--with-zlib'
```
Comment (lindsayad) [inline scripts/configure_petsc.sh:76]:
```suggestion
    HDF5_CONFIGURE_STR='--download-hdf5-configure-arguments="--with-zlib"'
```
This must have been a mistake in my commit

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: scripts/configure_petsc.sh (hunk 2 above)
Comment (cticenhour) [reply scripts/configure_petsc.sh:76]: Will fixup the original commit.

**Review by lindsayad — APPROVED** [review]

**Comment by cticenhour** [comment]
Thanks @lindsayad! Apptainer build cuda is a known failure, and has temporarily been disabled in idaholab/civet_recipes#1457. Other failures are unrelated or allowed to fail pending independent fixup from application leads. 

@loganharbour - re-requesting a review from you, pending the completion of the CI package builds.

**Comment by loganharbour** [comment]
Running them now. We don't need them to be complete, I just want a sanity check on if one build works :D

File: conda/petsc/build.sh [hunk 3]
```diff
@@ -36,18 +36,19 @@ else
...
     --CXXOPTFLAGS=-O3 \
     --FOPTFLAGS=-O3 \
     --with-x=0 \
     --with-ssl=0 \
-    CC="$CC" \
-    CXX="$CXX" \
-    FC="$FC" \
-    F90="$F90" \
-    F77="$F77" \
+    --with-mpi-dir=$PREFIX \
```
Comment (loganharbour) [inline conda/petsc/build.sh:51]: How does this work?

Comment (cticenhour) [reply conda/petsc/build.sh:51]: MPI (via `moose-mpich`) is installed in the `$PREFIX` location as defined by `conda-build`. This is the temporary environment that build creates and then PETSC is also installed to (see `--prefix=$PREFIX` a few lines down in the same file). Then packaging occurs and generalizes this. This only impacts the build and installation - conda takes care of the rest. 

Does that make sense?

Comment (loganharbour) [reply conda/petsc/build.sh:51]: oh, absolutely. I would have gotten that had I looked at the path... whoops

**Review by loganharbour — APPROVED** [review]
Good to go. One of them built, which is good enough for me. CUDA container is related, but ignored for now. We've decided to stop building this container as its use is null at the moment, it needs more work, and this is more important.

File: conda/petsc/build.sh (hunk 3 above)
Comment (cticenhour) [reply conda/petsc/build.sh:51]: Using $PREFIX here still maintains consistency with the end-user environment. This PETSc will always use the MPI provided by conda.

**Pull request merged**